toc/rfc/rfc-draft-add-support-for-file-based-service-binding.md

```diff
@@ -0,0 +1,110 @@
+- Updates of the Service Binding Information require restage. This is not optimal for an eventual support of [Binding rotation](https://github.com/openservicebrokerapi/servicebroker/blob/master/spec.md#binding-rotation) specification from the OSBI API spec.
+
+
+## Proposal
+
+The CFF community should implement an alternative approach for service binding information based on `tmpfs` file(s). Using a file or files to provide service binding information to applications will address the challenges listed in the problem section because:
+- The file size limit can be controlled by the CF platform
+- Already today CF uses `tmpfs` for [Instance Identity Credentials](https://docs.cloudfoundry.org/devguide/deploy-apps/instance-identity.html) which are rotated without restarting the application every `24h` by default
+
+The two approaches should be supported in parallel. Users should be able to select which approach Cloud Controller should use to deliver the binding information. Applications with binding information >`130KB` have to go with the file option and adopt it. There are two alternatives regarding service binding file organization:
+1. The VCAP_SERVICES content is stored in a file which location is specified via the VCAP_SERVICES_FILE_PATH env var in the same format as the VCAP_SERVICES environment variable
+   * Advantages:
+      * Less disruptive for applications consuming the `VCAP_SERVICES` env var
+      * Less implementation effort for the Cloud Controller
+   * Disadvantages:
+      * Can’t make use of tools and libraries from the Cloud Native community because K8s specifies a different file structure and format for the service binding information
+2. Implement the K8s service binding specification. The environment variable `SERVICE_BINDING_ROOT` defines the location for the service bindings. There is a file per service binding. The name of the file and the format follow the [K8s specification](https://servicebinding.io/):
+   * Advantages:
+      * CF community could re-use service binding libraries from the Cloud Native community
+      * Moving application between CF or K8s deployments will be easier
+   * Disadvantages
+      * Higher implementation efforts for the Cloud Controller
+
+
+The 2) alternatives offers more than just addressing the issue of this RFC. It suggests an option to evolve the CF platform towards a different service binding specification defined by the Cloud Native community. This means higher implementation efforts for the CF platform and application developers, but possible benefits from the Cloud Native community. This RFC has a light preference for the 2) alternative because of the listed advantages but the feedback of the CF community is wanted here.
```

Comment:
Some app developers are going to want to make the minimum changes possible, and 1 definitely works there. Other app developers are going to porting an app from k8s, and 2 makes their path onto CF easier. Is it possible to offer **both** 1 and 2, and have developers pick which environment variable they'll use to find the file?

Reply:
My read is that users will be able to pick between using the existing env variable approach, which means zero change for their apps, or they can use the file-based approach which would mean some level of change for their app with either approach. So there is control for the user that doesn't want change.

Both of the implementations above sound like there would be work involved. Minimally, you're going to need to know to read from a file instead. It's small, but if you've got apps that are not actively being developed, even that may be more work than you're willing to do.

Not sure if it needs to go into this proposal, but I would suggest that we target the common libraries that folks are using to read `VCAP_SERVICES`, like `java-cfenv` (I believe there are Python & Node.js libraries also) and get support for both formats, env variable and file, and then it's a lot less relevant how it's implemented. The libraries will smooth things over and possibly even make it so that no code change is required at the app level.

Reply:
>My read is that users will be able to pick between using the existing env variable approach, which means zero change for their apps, or they can use the file-based approach which would mean some level of change for their app with either approach. So there is control for the user that doesn't want change.

This is correct. @mogul given this clarification do you still see benefits in supporting both options in parallel? 

>Not sure if it needs to go into this proposal, but I would suggest that we target the common libraries that folks are using to read VCAP_SERVICES, like java-cfenv (I believe there are Python & Node.js libraries also) and get support for both formats, env variable and file, and then it's a lot less relevant how it's implemented. The libraries will smooth things over and possibly even make it so that no code change is required at the app level.

@dmikusa Do you mean here option 1) or both? Most probably both but I wanted to check?

Reply:
Both. I think regardless of what we do, we can lean on the libraries to help with smoothing things over for hopefully a lot of users.

Reply:
Just to comment on the reason for both options: There is a problem with the length of VCAP and running with existing VCAP + K8s service binding would not simplify to use the libraries to get the variables. So, to run both alternatives in parallel, I would propose to store the variables in both formats in two different files, not only keep the current VCAP in memory. This would solve the existing problem and also prepare for future service bindings.

Reply:
>  @mogul given this clarification do you still see benefits in supporting both options in parallel?

Nope, I hadn't taken into consideration that `VCAP_SERVICES` wasn't going away. I think it's OK to say "if you want better credential rotation or a larger set of bindings, then you have to use this new method which will require understanding `SERVICE_BINDING_ROOT`".

Reply:
If a user is hitting the length limit of `VCAP_SERVICES`, I would make the case that the user should then switch to the CNB bindings option which wouldn't suffer from that problem.

If you're using a library to parse `VCAP_SERVICES`, the idea would be that you can toggle between the two implementations (VCAP ENV or CNB) and your app wouldn't notice the difference. That should be achievable, at least with the Java/Spring libraries I've worked on.

Even if a user is not using a library and can't for some reason, both proposed implementations will require changes to the user's app, 1.) to read the file name from env variable, then read VCAP contents from a file, or 2.) read binding info from CNB binding files. Again, why not just read in the CNB bindings?

For those unfamiliar with the bindings spec, implementing it in a client only requires reading a couple of files from a directory. It's straightforward to find and read dependency data. I think even apps that have profile scripts that might `jq` some info out of `VCAP_SERVICES` should be able to switch and do something similar with a bit of bash script (it might not be a one-liner, but maybe two or three lines).

Reply:
> There is a problem with the length of VCAP and running with existing VCAP + K8s service binding would not simplify to use the libraries to get the variables. 

If their logic was to look for `SERVICE_BINDING_ROOT` first, and only fallback to reading from `VCAP_SERVICES` if it's not present, I think libraries would simplify this.

Reply:
I fully agree with the last commend provided by @dmikusa:
> If a user is hitting the length...

Reply:
I would like to resolve this so that the RFC could be updated with the option(s) we want to implement. That is why please vote:

| Option 1    | Option 2 | Both | Abstain |
| -------- | ------- |------- |------- |
| :+1:  | :thumbsdown:  | :rocket: | :eyes: |

cc TOC: @stephanme, @rkoster, @ChrisMcGowan, @ameowlia

Reply:
Should we consider :one:, :two:, :three: and :four: as kinda neutral alternatives to the vote indicators you proposed. No strong opinion, but e.g. 👎 seems to have a negative connotation and 🚀 seems to signal a very clear preference...

;)

Reply:
Ok, let's do this. I thought about this but also wanted to make it easier without a need of commenting :-)

Reply:
Just to clarify that the voting is open also for others not only for the TOC.

Reply:
:two:

Reply:
2️⃣

Reply:
2️⃣

Reply:
2️⃣

Reply:
2️⃣

Reply:
2️⃣

Reply:
:two:

Reply:
2️⃣

Reply:
2️⃣

Reply:
Thank you for the votes! I didn't set an end date for the voting but most of the RFC reviewers and the TOC already voted and the winner is clear. That is why, I updated the RFC (https://github.com/cloudfoundry/community/pull/804/commits/88f5fb6e54b3c292396d5b56157821d5b7133e04) with the winning option. Please review the latest state.